### ISSUE TITLE
Training Dataset Tags - Not Showing

### ISSUE THREAD

**Issue by LindezaBlue** [ostris/ai-toolkit#624]
## This is for bugs only

Did you already ask [in the discord](https://discord.gg/VXmU2f5WEU)?

No (Don't use Discord)

You verified that this is a bug and not a feature request or question by asking [in the discord](https://discord.gg/VXmU2f5WEU)?

No Same as above

## Describe the bug

After a successful training run (models work correctly) in the info pane there Training Dataset Tags are not present.
I have tried adding kohyaSS scripts/code to force it to show Training Dataset Tags, like:

**save_metadata:** true
**save_caption_metadata:** true
**save_dataset_metadata:** true
**store_dataset_info:** true
**write_tag_frequency:** true

Nothing seems to work, is this not a feature in AI-Toolkit? Or am I just missing something so simple and obvious? Any help would be appreciated. Thank you!~

## Example of what it should look like:
<img width="882" height="157" alt="Image" src="https://github.com/user-attachments/assets/0e68a7e9-284c-464a-9c6a-602214715fee" />

## What it actually looks like:
<img width="885" height="116" alt="Image" src="https://github.com/user-attachments/assets/82747091-83fc-46d3-82d6-fcf3155b56d6" />

## My Configuration for Training:
```job: "extension"
config:
  name: "Charact3r (illust) v2b"
  process:
    - type: "diffusion_trainer"
      training_folder: "E:\\PC BACKUP\\ai-toolkit\\output"
      sqlite_db_path: "./aitk_db.db"
      device: "cuda"
      trigger_word: charact3r
      performance_log_every: 10
      network:
        type: "lora"
        linear: 24
        linear_alpha: 24
        conv: 12
        conv_alpha: 12
        lokr_full_rank: true
        lokr_factor: -1
        network_kwargs:
          ignore_if_contains: []
      save:
        dtype: "bf16"
        save_every: 450
        max_step_saves_to_keep: 4
        save_format: "safetensors"
        save_metadata: true
        push_to_hub: false
      datasets:
        - folder_path: "E:\\PC BACKUP\\ai-toolkit\\datasets/charact3r"
          mask_path: null
          mask_min_value: 0.1
          default_caption: ""
          caption_ext: "txt"
          caption_dropout_rate: 0.01
          cache_latents_to_disk: true
          is_reg: false
          network_weight: 1
          resolution:
            - 1024
          controls: []
          shrink_video_to_frames: true
          num_frames: 1
          do_i2v: false
          flip_x: false
          flip_y: false
      train:
        batch_size: 1
        bypass_guidance_embedding: false
        steps: 450
        gradient_accumulation: 1
        train_unet: true
        train_text_encoder: true
        gradient_checkpointing: true
        noise_scheduler: "ddpm"
        optimizer: "adamw8bit"
        timestep_type: "sigmoid"
        content_or_style: "balanced"
        optimizer_params:
          weight_decay: 0.04
        unload_text_encoder: false
        cache_text_embeddings: false
        lr: 0.0001
        ema_config:
          use_ema: false
          ema_decay: 0.99
        skip_first_sample: false
        force_first_sample: false
        disable_sampling: true
        dtype: "bf16"
        diff_output_preservation: false
        diff_output_preservation_multiplier: 1
        diff_output_preservation_class: "person"
        switch_boundary_every: 1
        loss_type: "mse"
        ema_config?:
          ema_decay: 0.99
      logging:
        log_every: 1
        use_ui_logger: true
      model:
        name_or_path: "E:\\PC BACKUP\\ai-toolkit\\base_models\\illustriousXL_v01.safetensors"
        quantize: false
        qtype: "qfloat8"
        quantize_te: false
        qtype_te: "qfloat8"
        arch: "sdxl"
        low_vram: false
        model_kwargs: {}
      sample:
        sampler: "ddpm"
        sample_every: 100
        width: 1024
        height: 1024
        samples: []
        neg: ""
        seed: 42
        walk_seed: true
        guidance_scale: 5.5
        sample_steps: 25
        num_frames: 1
        fps: 1
meta:
  name: "Trained Model Test"
  version: "1.0"```

**Comment by LindezaBlue** [comment]
<p align="center">
  <img src="https://github.com/user-attachments/assets/faf10b44-ef58-4071-bc71-e85cb662499e" width="30%" alt="Program Interface">
</p>

## Why I Created This Tool

When training LoRAs with AI-Toolkit, I discovered that the resulting `.safetensors` files were missing crucial metadata that tools like Automatic1111 and Kohya SS rely on for features like automatic tag completion and dataset statistics display. Specifically, the `ss_tag_frequency` and related metadata fields weren't being embedded into the trained models.

After opening [an issue (#624)](https://github.com/ostris/ai-toolkit/issues/624) and waiting for community feedback on how to add this metadata natively within AI-Toolkit, I realized there wasn't a straightforward solution available. *(That and help is mostly on Discord which I don't use, and shouldn't have to use to get help.)* 

Rather than manually editing metadata or waiting for a native implementation, I decided to create a standalone tool that could inject this metadata directly into existing LoRA files.

This tool:
- ✅ Scans your training dataset and calculates tag frequencies
- ✅ Embeds Kohya/A1111-compatible metadata into your LoRA files
- ✅ Works retroactively on already-trained models
- ✅ Requires no modifications to AI-Toolkit itself

### 🔗 Repository
**[Dataset Metadata Injection Tool](https://github.com/LindezaBlue/Dataset-Metadata-Injection)**

---

Hopefully this helps anyone else who's encountered the same issue with AI-Toolkit trained LoRAs! Feel free to let me know if this helped you as well!

**Comment by Jacid23** [comment]
AI-Toolkit doesn't write training metadata to your safetensors files. Out of the box, the only thing it saves is the trigger word with a count of 1. That's why A1111/Forge/ComfyUI don't show trigger words, tags, or training params on AI-Toolkit LoRAs like they do for Kohya-trained ones.

There's no setting to turn this on — the code simply doesn't write it. Here's how to add it.

## The Fix

**File to edit:**
```
ai-toolkit/jobs/process/BaseSDTrainProcess.py
```

> Back this up first. Updates will overwrite it.

### Step 1 — Add imports

At the top of the file, make sure `Counter` and `json` are imported:

```python
from collections import OrderedDict, Counter
import json
import os
```

`OrderedDict` should already be there. Just add `Counter` next to it and add the `json` and `os` imports if missing.

### Step 2 — Replace `update_training_metadata()`

Find the `update_training_metadata` method. Stock version only writes the trigger word. Replace the entire method with:

```python
def update_training_metadata(self):
    o_dict = OrderedDict({
        "training_info": self.get_training_info()
    })
    o_dict['ss_base_model_version'] = self.sd.get_base_model_version()
    o_dict['ss_output_name'] = self.job.name

    # --- Network config (LoRA rank, alpha, type) ---
    if self.network_config is not None:
        o_dict['ss_network_module'] = self.network_config.type
        o_dict['ss_network_dim'] = self.network_config.linear
        o_dict['ss_network_alpha'] = self.network_config.linear_alpha
        if self.network_config.conv is not None:
            o_dict['ss_network_dim_conv'] = self.network_config.conv
            o_dict['ss_network_alpha_conv'] = self.network_config.conv_alpha
        if self.network_config.dropout is not None:
            o_dict['ss_network_dropout'] = self.network_config.dropout

    # --- Training config ---
    o_dict['ss_learning_rate'] = self.train_config.lr
    o_dict['ss_unet_lr'] = self.train_config.unet_lr
    o_dict['ss_text_encoder_lr'] = self.train_config.text_encoder_lr
    o_dict['ss_optimizer'] = self.train_config.optimizer
    o_dict['ss_lr_scheduler'] = self.train_config.lr_scheduler
    o_dict['ss_max_train_steps'] = self.train_config.steps
    o_dict['ss_batch_size_per_device'] = self.train_config.batch_size
    o_dict['ss_mixed_precision'] = self.train_config.dtype
    o_dict['ss_gradient_checkpointing'] = self.train_config.gradient_checkpointing
    o_dict['ss_noise_offset'] = self.train_config.noise_offset
    if self.train_config.min_snr_gamma is not None:
        o_dict['ss_min_snr_gamma'] = self.train_config.min_snr_gamma
    o_dict['ss_seed'] = self.train_config.seed if hasattr(self.train_config, 'seed') else None
    if hasattr(self.train_config, 'gradient_accumulation_steps'):
        o_dict['ss_gradient_accumulation_steps'] = self.train_config.gradient_accumulation_steps

    # --- Model config ---
    o_dict['ss_sd_model_name'] = self.model_config.name_or_path
    if self.model_config.is_flux:
        o_dict['ss_model_arch'] = 'flux'
    elif self.model_config.is_xl:
        o_dict['ss_model_arch'] = 'sdxl'
    elif self.model_config.is_v2:
        o_dict['ss_model_arch'] = 'sd2'
    else:
        o_dict['ss_model_arch'] = 'sd1'

    # --- Current step/epoch ---
    o_dict['ss_steps'] = self.step_num
    o_dict['ss_epoch'] = self.epoch_num

    # --- Real tag frequencies from caption files ---
    tag_frequency = OrderedDict()
    dataset_dirs = OrderedDict()
    total_images = 0

    if self.dataset_configs:
        for ds_config in self.dataset_configs:
            if ds_config.is_reg:
                continue
            folder = ds_config.folder_path or ds_config.dataset_path
            if not folder:
                continue

            folder_name = os.path.basename(folder.rstrip('/\\'))
            n_repeats = ds_config.num_repeats if hasattr(ds_config, 'num_repeats') else 1
            dir_key = f"{n_repeats}_{folder_name}"

            caption_ext = ds_config.caption_ext or '.txt'
            ds_tags, img_count = self._scan_caption_tags(folder, caption_ext)

            if img_count > 0:
                tag_frequency[dir_key] = ds_tags
                dataset_dirs[dir_key] = {
                    "n_repeats": n_repeats,
                    "img_count": img_count,
                }
                total_images += img_count

            # Resolution
            if hasattr(ds_config, 'resolution') and ds_config.resolution:
                res = ds_config.resolution
                if isinstance(res, (list, tuple)):
                    o_dict['ss_resolution'] = ','.join(str(x) for x in res)
                else:
                    o_dict['ss_resolution'] = f"{res},{res}"

    if tag_frequency:
        o_dict['ss_tag_frequency'] = tag_frequency
        o_dict['ss_dataset_dirs'] = dataset_dirs
        o_dict['ss_num_train_images'] = total_images
    elif self.trigger_word is not None:
        o_dict['ss_tag_frequency'] = {
            f"1_{self.trigger_word}": {self.trigger_word: 1}
        }

    if self.trigger_word is not None:
        o_dict['ss_trigger_word'] = self.trigger_word

    if self.embed_config is not None:
        o_dict['ss_embedding_trigger'] = self.embed_config.trigger
        o_dict['ss_embedding_tokens'] = self.embed_config.tokens

    if self.train_config.optimizer_params:
        o_dict['ss_optimizer_params'] = self.train_config.optimizer_params
    if self.train_config.lr_scheduler_params:
        o_dict['ss_lr_scheduler_params'] = self.train_config.lr_scheduler_params

    # Remove None values
    o_dict = OrderedDict({k: v for k, v in o_dict.items() if v is not None})
    self.add_meta(o_dict)
```

### Step 3 — Add the tag scanner

Right below `update_training_metadata`, add this static method:

```python
@staticmethod
def _scan_caption_tags(folder_path, caption_ext='.txt'):
    """Scan caption files and return tag frequencies + image count."""
    tag_counter = Counter()
    image_count = 0
    img_exts = {'.png', '.jpg', '.jpeg', '.webp', '.bmp', '.tiff'}

    if not caption_ext.startswith('.'):
        caption_ext = '.' + caption_ext

    if not os.path.isdir(folder_path):
        return dict(tag_counter), image_count

    for fname in os.listdir(folder_path):
        fpath = os.path.join(folder_path, fname)
        if not os.path.isfile(fpath):
            continue
        ext = os.path.splitext(fname)[1].lower()
        if ext in img_exts:
            image_count += 1
            cap_file = os.path.join(
                folder_path, os.path.splitext(fname)[0] + caption_ext
            )
            if os.path.isfile(cap_file):
                try:
                    with open(cap_file, 'r', encoding='utf-8') as f:
                        text = f.read().strip()
                    for tag in text.split(','):
                        tag = tag.strip()
                        if tag:
                            tag_counter[tag] += 1
                except Exception:
                    pass

    return dict(tag_counter), image_count
```

## What You Get

After this, every LoRA checkpoint will contain Kohya-compatible metadata:

| Key | What it is |
|-----|-----------|
| ss_tag_frequency | Real tag counts from your captions — A1111/Forge shows these as trigger words |
| ss_dataset_dirs | Folder names, image counts, repeats |
| ss_network_module/dim/alpha | LoRA type, rank, alpha |
| ss_learning_rate | LR used |
| ss_optimizer | Optimizer name |
| ss_sd_model_name | Base model |
| ss_model_arch | flux / sdxl / sd1 / sd2 |
| ss_resolution | Training resolution(s) |
| ss_steps | Step number at save |
| ss_batch_size_per_device | Batch size |
| ss_mixed_precision | Training dtype |

No config changes needed — once patched, it writes automatically on every save.

## Heads Up

- **Updates overwrite this file.** Back it up or you'll need to re-apply after updating AI-Toolkit.
- This reads `.txt` caption files. If you use a different caption format, the tag scanning won't pick them up (everything else still works).
- This is the same metadata format Kohya uses — all existing metadata readers work with it.
- It doesn't affect training at all. It's just extra info baked into the file.

**Comment by Jacid23** [comment]
sorry I kinda killed all your work with your tool. FYI I forked your tool, to add additional metadata that's when light bulb went off to have Claude dig through the files to see if we can just add it make life simple and it was all there just off. I hate apps that are so dumbed down that makes it difficult for a mid level person to use it. There are tons of options in there that unless you write code you will never know. I had Claude write me up a "Real" Flux training config.yaml with all possible setting as well as options. He created a mimic format of their example with #to comment out so its easy edit. AItoolkit = 96 lines of code vs Claude Opus 4.6 = 600 yes 600 lines of code 30 of which are just separations to make it easier on the eyes, as well as a ton of stuff you would never touch BUT there is def a good deal of things, coming for kohya_ss I was wow all the settings are here. Also note flowmatch = "flux euler" aka euler I guess SD version is a bit different but most tools handle that in backend without us knowing that the euler for flux is different, it does technically the same thing just for flux. Well hope this save you time from adding to app or being forced to use it.

**Comment by LindezaBlue** [comment]
Not sure what you mean by "killed all my work."

My tool still works for adding missing metadata for any model even ones
from models not trained for OneTrainer etc.

<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by Jacid23** [comment]
Ok if you like to do extra steps, myself I much rather let the trainer automatically add all the metadata.
